sys/net/gnrc/network_layer/ipv6/nib/_nib-router.c

```diff
@@ -228,7 +264,9 @@ static void _snd_ra(gnrc_netif_t *netif, const ipv6_addr_t *dst,
 {
     gnrc_pktsnip_t *ext_opts = NULL;
 
-    if (!final) {
+    if (final) {
+        ext_opts = _build_final_ext_opts(netif);
+    } else {
```

Comment:
Ok, now let's discuss this... Why are the other options not added? Why is the RIO now only included in the final advertisement? The final router advertisement basically tells the surrounding nodes "Do not consider me a router anymore" but with this option you say "but do route that traffic via me". Can you give a reference for that?

Reply:
That's the thing that confused me too.
A border router will sent this 'final RA', but it's still a router, so my interpretation was that this can't signal "I'm not a router".
And when I added the RIO, Linux would just do the right thing and create a

```
2001:16b8:4545:14f8::/62 via fe80::a4ed:29ff:feec:cbf3 dev eno1 proto ra metric 100 pref medium
```

so I was happy.

Reply:
Ok, yeah, a router that does not self-advertise as a router is basically just not a default router, but why then add the RIO _only_ in the final RA?

Reply:
If we advertise ourselves as a default router, we expect *everything* to be routed via us anyway.
So always adding the RIO will add a superfluous entry to the forwarding table on the downstream node:

```
2021-06-28 14:58:17,217 # 2001:16b8:4545:1400::/64 via fe80::a4ed:29ff:feec:d5f3 dev #7 // this is not needed
2021-06-28 14:58:17,220 # 2001:16b8:4545:14f8::/62 dev #7
2021-06-28 14:58:17,220 # 2001:16b8:4545:14fa::/63 dev #6
2021-06-28 14:58:17,222 # default via fe80::a4ed:29ff:feec:d5f3 dev #7
```

Reply:
> If we advertise ourselves as a default router, we expect _everything_ to be routed via us anyway.

No, only the packets for which no other route can be found. ;-).



> So always adding the RIO will add a superfluous entry to the forwarding table on the downstream node:
> 
> ```
> 2021-06-28 14:58:17,217 # 2001:16b8:4545:1400::/64 via fe80::a4ed:29ff:feec:d5f3 dev #7 // this is not needed
> 2021-06-28 14:58:17,220 # 2001:16b8:4545:14f8::/62 dev #7
> 2021-06-28 14:58:17,220 # 2001:16b8:4545:14fa::/63 dev #6
> 2021-06-28 14:58:17,222 # default via fe80::a4ed:29ff:feec:d5f3 dev #7
> ```
> 
> It doesn't hurt, but takes up some space both in RAM and on the wire.

I thought this was the whole plan? If not and you only want the border router to advertise their downstream route, why not only advertise add the RIO only in the upstream router advertisements of the border router. Much less confusing than tying it to the final router advertisement (which can also happen on a downstream node if the `rtr_adv` or `rtr` flags is unset for whatever reason.

Reply:
Hm, are you suggesting to add a new flag e.g. `GNRC_NETIF_FLAGS_IPV6_RTR_ADV_RIO_ONLY` that we then can set on the upstream interface?

Reply:
Why RIO-only? Wouldn't just `GNRC_NETIF_FLAGS_IPV6_RTR_ADV_RIO` be enough?

Reply:
I think [this comment](https://github.com/RIOT-OS/RIOT/pull/16568#discussion_r659545433) still applies.
We certainly wouldn't want to add the ABR option when advertising to the upstream network.
None of the options in `_build_ext_opts` make sense when sending a RA upstream - we just got them from there, no need to echo them back.

Reply:
> I think [this comment](https://github.com/RIOT-OS/RIOT/pull/16568#discussion_r659545433) still applies.

None of the items of that comment are tied to any options. The default route comes from the router advertisement itself and sending router solicitations is basically the opposite of sending router advertisements.

> We certainly wouldn't want to add the ABR option when advertising to the upstream network.

If that happens, we certainly have a bug anyways. The ABR option should only go out the 6LoWPAN border router (read downstream) interface.

> None of the options in `_build_ext_opts` make sense when sending a RA upstream - we just got them from there, no need to echo them back.

Except that adding an exception to the rule adds costly memory. It's not like the packet size for the upstream route needs to be kept small either, it is most of the time some Ethernet/Wifi link. Finally, it just smells like a bad or at least arbitrary design decision. Why does only the RIO get this `only` option? Why is there no way to send a RIO with the other options? Someone reading this code later down the line will ask these questions and also why this decisions were made. Since you are touching very central IPv6 code and going "of spec" here, your answer should be very concise and well reasoned.

Reply:
Ok, I added the `GNRC_NETIF_FLAGS_IPV6_RTR_ADV_RIO` flag and updated the last commit accordingly.

Reply:
Hmm so this does not work. If `rtr_adv->ltime.u16 != 0` we try to add the node as a default router in `_handle_rtr_adv()`.
This will fail for the default `GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF=1` (and setting this to two appears to cause different issues).

In the 'last RA' case we set `ltime` to 0, this causes the receiver to *not* add it to the default router list, but still processes all the options.

Reply:
So how about hiding this behind `CONFIG_GNRC_IPV6_NIB_RIO_IN_LAST_RA`?